verify.js

```diff
@@ -86,96 +51,143 @@ module.exports = function (jwtString, secretOrPublicKey, options, callback) {
 
   var header = decodedToken.header;
 
-  if (!~options.algorithms.indexOf(header.alg)) {
-    return done(new JsonWebTokenError('invalid algorithm'));
+  if(header.alg === 'none' && decodedToken.signature) {
+    return done(new JsonWebTokenError('invalid token: unsigned but signature is present'));
   }
 
-  var valid;
+  var getSecret;
 
-  try {
-    valid = jws.verify(jwtString, header.alg, secretOrPublicKey);
-  } catch (e) {
-    return done(e);
+  if(typeof secretOrPublicKey !== 'function') {
+    getSecret = function(header, callback) {
+      return callback(null, secretOrPublicKey);
```

Comment:
Please don't shadow the `callback` variable from the outer scope in the closure, it makes it confusing to read.